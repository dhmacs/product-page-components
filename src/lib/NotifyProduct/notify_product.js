import React, { useState } from 'react';
import styled from 'styled-components';
import { _js } from '@ifixit/localize';
import { Button, TextField } from '@ifixit/toolbox';
import { color } from '@core-ds/primitives';
import Status, { StatusType } from './status';
import { post } from '../api';

/**
 * Because we have three components with different heights that swap their
 * displays on and off, we need to hard-code the container's height to be the
 * same as the height of the biggest component (in this case the confirmation)
 * in order to keep the overall size constant.
 */
const NotifyContainer = styled.div`
   height: 56px;
`;

const InitialBanner = styled.span`
   justify-content: space-between;

   ${({ visible }) =>
      visible
         ? `
      display: flex;
   `
         : `
      display: none;
   `};
`;

const OutOfStock = styled.span`
   display: flex;
   align-items: center;
   color: ${color.red};
`;

const EmailForm = styled.form`
   ${({ visible }) =>
      visible
         ? `
      display: flex;
   `
         : `
      display: none;
   `}

   .text-field {
      padding-right: 10px;
   }
`;

const HidableStatus = styled(Status)`
   ${({ visible }) =>
      visible
         ? `
      display: block;
   `
         : `
      display: none;
   `}
`;

const notifyStage = {
   INITIAL: 'initial',
   INPUT: 'input',
   CONFIRMATION: 'confirmation',
};

const defaultConfirmationStatus = {
   successful: true,
   message: "We'll send a message when the product is back in stock.",
};

const NotifyProduct = ({ email, productcode, optionid }) => {
   const [stage, setStage] = useState(notifyStage.INITIAL);
   const [confirmationStatus, setConfirmationStatus] = useState(defaultConfirmationStatus);
   const [formEmail, setFormEmail] = useState(email);

   return (
      <NotifyContainer>
         <InitialBanner visible={stage === notifyStage.INITIAL}>
            <Button design="outline" onClick={() => setStage(notifyStage.INPUT)}>
               Notify me
            </Button>
            <OutOfStock>Out of Stock</OutOfStock>
         </InitialBanner>
         <EmailForm visible={stage === notifyStage.INPUT}>
            <TextField
               showValidity
               required
               type="email"
               className="text-field"
               value={formEmail}
               placeholder="Enter your email"
               onChange={({ value }) => setFormEmail(value)}
            />
            <Button
               design="primary"
               disabled={!formEmail}
               onClick={event => {
                  event.preventDefault();
                  post('cart/product/notifyWhenInStock', {
                     productcode,
                     optionid,
                     email: formEmail,
                  })
                     .then(() => setStage(notifyStage.CONFIRMATION))
                     .catch(err => {
                        setStage(notifyStage.CONFIRMATION);
                        setConfirmationStatus({
                           successful: false,
                           message: err.message,
                        });
                     });
               }}
            >
               Notify me
            </Button>
         </EmailForm>
         <HidableStatus
            visible={stage === notifyStage.CONFIRMATION}
            type={confirmationStatus.successful ? StatusType.SUCCESS : StatusType.ERROR}
         >
            {confirmationStatus.message}
         </HidableStatus>
      </NotifyContainer>
   );
};

export default NotifyProduct;
