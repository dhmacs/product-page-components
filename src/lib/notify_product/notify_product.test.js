import React from 'react';
import NotifyProduct from './notify_product';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

it('renders correctly', () => {
   const tree = renderer
      .create(<NotifyProduct productcode={145307} optionid={4} salesChannelID={1} />)
      .toJSON();
   expect(tree).toMatchSnapshot();
});
