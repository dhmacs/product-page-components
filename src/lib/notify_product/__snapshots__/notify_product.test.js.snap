// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  position: relative;
  height: 63px;
  overflow: hidden;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-transition: opacity 0.5s;
  transition: opacity 0.5s;
  visibility: visible;
  opacity: 1;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #F34E46;
  font-size: 14px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-transition: all 0.5s;
  transition: all 0.5s;
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
  visibility: hidden;
  opacity: 1;
}

.c3 .text-field {
  padding-right: 10px;
}

.c3 .text-field > span {
  display: none;
}

.c5 {
  margin: 0;
  border-width: 1px;
  border-style: solid;
  border-radius: 4px;
  padding: 12px 12px;
  font-size: 13px;
  color: #288c4e;
  background-color: #ebf8f0;
  border-color: #99dab2;
  position: absolute;
  width: 100%;
  box-sizing: border-box;
  -webkit-transition: -webkit-transform 0.5s;
  -webkit-transition: transform 0.5s;
  transition: transform 0.5s;
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
  visibility: hidden;
}

.c4 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
}

<div
  className="c0"
>
  <span
    className="c1"
  >
    <button
      className="css-174r4qz"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
    >
      <span
        className="css-57kgtb"
      >
        <span
          className="css-tz46pd"
        >
          Notify me
        </span>
      </span>
    </button>
    <span
      className="c2"
    >
      Out of Stock
    </span>
  </span>
  <form
    className="c3"
  >
    <label
      className="css-n946dq text-field"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      
      <div
        className="css-13azwyo"
      >
        <div
          className="css-1eca8pd"
          onClick={[Function]}
        >
          <span
            className="css-1jr2kwt"
          >
            <span
              className="css-1oyzn7j"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"inherit\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"feather feather-check-circle\\"><path d=\\"M22 11.08V12a10 10 0 1 1-5.93-9.14\\"></path><polyline points=\\"22 4 12 14.01 9 11.01\\"></polyline></svg>",
                }
              }
            />
          </span>
        </div>
        <input
          className="css-1l6rqr7"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInvalid={[Function]}
          placeholder="Enter your email"
          required={true}
          type="email"
          value=""
        />
      </div>
    </label>
    <button
      className="css-1er3yna c4"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
    >
      <span
        className="css-57kgtb"
      >
        <span
          className="css-tz46pd"
        >
          Notify me
        </span>
      </span>
    </button>
  </form>
  <p
    className="c5"
    type="success"
  >
    We'll send a message when the product is back in stock.
  </p>
</div>
`;
